From c815d8feb9fc5a5c9cbd4feeb6fef4c73447f414 Mon Sep 17 00:00:00 2001
From: Fam Zheng <famz@redhat.com>
Date: Mon, 4 Aug 2014 09:48:46 -0400
Subject: [virt] hyperv/storvsc: Fix a bug in handling VMBUS protocol version

Message-id: <1407145729-25017-5-git-send-email-famz@redhat.com>
Patchwork-id: 86955
O-Subject: [RHEL-7 PATCH 4/7] Drivers: scsi: storvsc: Fix a bug in handling VMBUS protocol version
Bugzilla: 1122317
RH-Acked-by: Jason Wang <jasowang@redhat.com>
RH-Acked-by: Paolo Bonzini <pbonzini@redhat.com>

From: K. Y. Srinivasan <kys@microsoft.com>

Based on the negotiated VMBUS protocol version, we adjust the size of the storage
protocol messages. The two sizes we currently handle are pre-win8 and post-win8.
In WS2012 R2, we are negotiating higher VMBUS protocol version than the win8
version. Make adjustments to correctly handle this.

Signed-off-by: K. Y. Srinivasan <kys@microsoft.com>
Reviewed-by: Hannes Reinecke <hare@suse.de>
Cc: <stable@vger.kernel.org>
Signed-off-by: Christoph Hellwig <hch@lst.de>
(cherry picked from commit adb6f9e1a8c6af1037232b59edb11277471537ea)
Signed-off-by: Fam Zheng <famz@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/storvsc_drv.c b/drivers/scsi/storvsc_drv.c
index d21c716..6f8d76e 100644
--- a/drivers/scsi/storvsc_drv.c
+++ b/drivers/scsi/storvsc_drv.c
@@ -1749,19 +1749,22 @@ static int storvsc_probe(struct hv_device *device,
   * set state to properly communicate with the host.
   */
 
- if (vmbus_proto_version == VERSION_WIN8) {
-  sense_buffer_size = POST_WIN7_STORVSC_SENSE_BUFFER_SIZE;
-  vmscsi_size_delta = 0;
-  vmstor_current_major = VMSTOR_WIN8_MAJOR;
-  vmstor_current_minor = VMSTOR_WIN8_MINOR;
- } else {
+ switch (vmbus_proto_version) {
+ case VERSION_WS2008:
+ case VERSION_WIN7:
   sense_buffer_size = PRE_WIN8_STORVSC_SENSE_BUFFER_SIZE;
   vmscsi_size_delta = sizeof(struct vmscsi_win8_extension);
   vmstor_current_major = VMSTOR_WIN7_MAJOR;
   vmstor_current_minor = VMSTOR_WIN7_MINOR;
+  break;
+ default:
+  sense_buffer_size = POST_WIN7_STORVSC_SENSE_BUFFER_SIZE;
+  vmscsi_size_delta = 0;
+  vmstor_current_major = VMSTOR_WIN8_MAJOR;
+  vmstor_current_minor = VMSTOR_WIN8_MINOR;
+  break;
  }
 
-
  if (dev_id->driver_data == SFC_GUID)
   scsi_driver.can_queue = (STORVSC_MAX_IO_REQUESTS *
       STORVSC_FC_MAX_TARGETS);
-- 
1.7.1